modules/contact/src/constraints/ApplyPenetrationConstraintLMMechanicalContact.C

```diff
@@ -0,0 +1,210 @@
+#include "PenetrationLocator.h"
+#include "PenetrationInfo.h"
+#include "SystemBase.h"
+#include "Assembly.h"
+
+#include "DualRealOps.h"
+
+using MetaPhysicL::DualNumber;
+
+registerMooseObject("ContactApp", ApplyPenetrationConstraintLMMechanicalContact);
+
+InputParameters
+ApplyPenetrationConstraintLMMechanicalContact::validParams()
+{
+  InputParameters params = NodeFaceConstraint::validParams();
+  params.set<bool>("use_displaced_mesh") = true;
+
+  params.addCoupledVar("disp_y", "The y displacement");
+  params.addCoupledVar("disp_z", "The z displacement");
+  params.addParam<Real>("c", 1, "Parameter for balancing the size of the gap and contact pressure");
+
+  params.addClassDescription(
+      "Implements the KKT conditions for normal contact using an NCP "
+      "function, in this case just the min function. This function enforces that either "
+      "the gap distance or the normal contact pressure (represented by the value of `variable`) is "
+      "zero.");
+  return params;
+}
+
+ApplyPenetrationConstraintLMMechanicalContact::ApplyPenetrationConstraintLMMechanicalContact(
+    const InputParameters & parameters)
+  : NodeFaceConstraint(parameters),
+    _c(getParam<Real>("c")),
+    _fe_problem(*getCheckedPointerParam<FEProblemBase *>("_fe_problem_base")),
+    _displaced_problem(_fe_problem.getDisplacedProblem().get()),
+    _residual_copy(_sys.residualGhosted())
+{
+}
+
+bool
+ApplyPenetrationConstraintLMMechanicalContact::shouldApply()
+{
+  const bool has_dofs = _current_node->n_dofs(_sys.number(), _var.number());
+  _dof_number = has_dofs ? _current_node->dof_number(_sys.number(), _var.number(), /*comp=*/0)
+                         : DofObject::invalid_id;
+  return has_dofs;
+}
+
+bool
+ApplyPenetrationConstraintLMMechanicalContact::overwriteSecondaryResidual()
+{
+  // If we didn't project the secondary node onto the primary surface, then gap has no meaning and
+  // we just want to drive our Lagrange Multiplier to zero. If we did, then our LM drives the
+  // residual if it is less than the gap (we're using min for our NCP function)
+  return (!_projection_successful) ||
+         (_u_secondary[_qp] < _node_to_integrated_gap.at(_current_node));
+}
+
+void
+ApplyPenetrationConstraintLMMechanicalContact::residualSetup()
+{
+  NodeFaceConstraint::residualSetup();
+  _node_to_integrated_gap.clear();
+}
+
+Real
+ApplyPenetrationConstraintLMMechanicalContact::computeQpSecondaryValue()
+{
+  return _u_secondary[_qp];
+}
+
+void
+ApplyPenetrationConstraintLMMechanicalContact::computeResidual()
+{
+  DenseVector<Number> & re = _assembly.residualBlock(_var.number());
+
+  _qp = 0;
+
+  _secondary_residual = re(0) = computeQpResidual(Moose::Secondary);
+  _secondary_residual_computed = true;
+}
+
+void
+ApplyPenetrationConstraintLMMechanicalContact::computeJacobian()
+{
+  _Kee.resize(1, 1);
+  _connected_dof_indices.clear();
+  _connected_dof_indices.push_back(_var.nodalDofIndex());
+
+  _qp = 0;
```

Comment:
I am failing to see how the Gauss point variable is changing (maybe it shouldn't) in the calls to the *Qp* methods to fill residuals and Jacobians.

Reply:
Yea `_qp` is kind of a misnomer here because we are not applying any quadrature rules. We are just evaluating secondary values at one location (the secondary node). (And for `NodeFaceConstraints` which work the the primary face, obviously we only work with one location over there as well...the location where the secondary node projects). But to have consistency with the moose data structures we need to index into our `MooseArrays` holding variable value data; in this case we know the data should have size 1, so `_qp` must be 0. Perhaps it would be more transparent if we didn't do anything with `_qp` and explicitly just indexed with 0?

Reply:
Thanks for the details! I get it now. I think it's fine how it is, just lack of familiarity on my side.